rosutil/symdebug.rkt

```diff
@@ -0,0 +1,58 @@
+#lang racket
+
+(require (prefix-in r: rosette/safe))
+
+; Utilities for inspecting symbolic expressions in Yosys module state
+
+(provide term-length analyze-terms wide-write)
+
+; helper function that returns flattened length of an arbitrarily nested list
+(define (flattened-length l)
+  (if (not (list? l))
+      1
+      (if (empty? l)
+          0
+          (+ (flattened-length (first l))
+             (flattened-length (rest l))))))
+
+; returns "length" of (the possibly symbolic) t, depending on type of t:
+; vector? returns sum of term-length of each element
+; [symbolic] expression? returns flattened-length of t
+; [symbolic] constant? returns 1
+; other (i.e., concrete value) returns 0
+(define (term-length t)
```

Comment:
I don't quite understand the spec of this function. Here are some examples where it doesn't match my intuition:

```racket
#lang rosette/safe

(require rosutil)

(define-symbolic* b boolean?)
(define-symbolic* n integer?)
(define v (vector (bv 0 5) n (bv 2 5)))

(define x (if b (list 1 2) "hi"))

(term-length #t)
; => 0

(term-length x)
; => 0

(term-length (list 1 2))
; => 0

(term-length "hi")
; => 0

(term-length v)
; => 1

(vector-set! v n b)
(term-length v)
; => 0
```

Have you tried using the Rosette [value browser](https://docs.racket-lang.org/rosette-guide/sec_utility-libs.html#%28part._.Value_.Browser_.Library%29)? It would be neat if our way of measuring value sizes had some sort of correspondence with how they are displayed there. Here's an example of what that could look like: https://github.com/anishathalye/rtl/tree/value-size (where the correspondence is that the number returned by the function is equal to the number of boxes shown in the value browser when the value is fully expanded).

Reply:
I haven't seen the value browser before! That seems really handy, so good to know about. I'm definitely not opposed to using something like value-size in your branch, that seems fairly reasonable. I do think one part of the spec I wanted (and this addresses some of your previous comments) is that fully concrete values would simply return 0, to differentiate them from symbolic constants/expressions. I get that this feels a little inconsistent with counting constants that are parts of expressions though.